src/model.py

```diff
@@ -805,46 +805,40 @@ def init_optimizer(self):
         """
+        other_components = self.find("other_components", default={
+            component: {"lr": self.learning_rate, "tasks": []}
+            if component not in self.subtasks else {}
+            for component in self.components
+        })
+        optimizer_name = self.find("optimizer_name", default=self.find("optimizer"))
+        bert_tuners = {  # subtasks to train bert on
+            task: self.find(f"bert_{task}", default=False)
+            for task in self.subtasks
+        }
+        task_lrs = { # subtask specific learning rates
+            task: self.find(task+"_learning_rate", default=self.find(task+"_lr")) 
+            for task in self.subtasks
+        }
+
         optimizers = {}
         schedulers = {}
 
         # single optimizer type for all subtasks for simplicity
-        optimizer = self.get_optimizer(
-            self.find("optimizer_name", default=self.find("optimizer"))
-        )
-
-        # check if task specific learning rates are provided
-        task_lrs = {
-            task: self.find(task+"_learning_rate", default=self.find(task+"_lr")) 
-            for task in self.subtasks
-        }
+        optimizer = self.get_optimizer(optimizer_name)
         
-        for task in self.subtasks:  # TODO make sure in self.components in others
+        for task in self.subtasks:
             lr = task_lrs[task] if task_lrs.get(task) is not None else self.learning_rate
 
-            for i, layer in enumerate(self.components[task]):
-                if i == 0:
-                    optimizers[task] = optimizer(
-                        self.components[task][layer].parameters(),
-                        lr=lr  # use main learning rate for bert training
-                    ) # TODO test other optimizers?
-                    
-                    if "shared" in self.components.keys():
-                        for layer in self.components["shared"]:
-                            optimizers[task].add_param_group(
-                                {"params": self.components["shared"][layer].parameters(), "lr":lr}
-                            )
-
-                    if self.find(f"bert_{task}", default=False) and self.finetune:
-                        optimizers[task].add_param_group(
-                            {"params": self.bert.parameters(), "lr":self.learning_rate}
-                        )
-
-                else:
-                    optimizers[task].add_param_group(
-                        {"params": self.components[task][layer].parameters(), "lr":lr}
-                    )
+            optimizers[task] = optimizer(
+                self.components[task].parameters(),
+                lr=lr 
+            )
 
+            if bert_tuners[task] and self.finetune:
+                optimizers[task].add_param_group(
+                    # use main learning rate for bert training
+                    {"params": self.bert.parameters(), "lr":self.learning_rate}
+                )
```

Comment:
Then `bert.parameters()` are added to those tasks allowed to train Bert.

Notice the `bert_tuners` dict defined at the top of the method defaults to `False` for all tasks. This dict looks for attributes of the form `bert_{task} (bool)` of the model.